### ISSUE TITLE
Internal server error response on production mode with MySQL server, but transaction was OK

### ISSUE THREAD

**Issue by avansp** [jhipster/generator-jhipster#5499]
My app always showed "error.internalServerError" message when I did any transactions (create/edit/delete) to a MySQL database server. Cancelled the dialog box and refreshed the table. The transaction was done perfectly. And this only happened in the production mode. Tested on dev mode, everything worked fine, no "internal server error" message. I don't know how to trace this bug.

Note that I tested the production mode using previous jhipster version, it didn't show this message, but I forgot which version it was.

### **Reproduce error**

To test if this wasn't my application, I generated a fresh jhipster app (see the info below). Used different database schema. The error message still showed up but the transaction was done.

Welcome to the JHipster Information Sub-Generator

##### **JHipster Version(s)**

```
mysql-test@0.0.0 /mysql_test
└── generator-jhipster@4.1.1 

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

```yaml
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "org.cardiacatlas.xpacs"
    },
    "jhipsterVersion": "4.1.1",
    "baseName": "mysql_test",
    "packageName": "org.cardiacatlas.xpacs",
    "packageFolder": "org/cardiacatlas/xpacs",
    "serverPort": "8080",
    "authenticationType": "session",
    "hibernateCache": "ehcache",
    "clusteredHttpSession": false,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "mysql",
    "searchEngine": "elasticsearch",
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "maven",
    "enableSocialSignIn": false,
    "rememberMeKey": "***",
    "clientFramework": "angular1",
    "useSass": false,
    "clientPackageManager": "yarn",
    "applicationType": "monolith",
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "enableTranslation": false
  }
}
```

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

PatientInfo.json

```yaml
{
    "fluentMethods": true,
    "relationships": [],
    "fields": [
        {
            "fieldName": "patient_id",
            "fieldType": "String",
            "fieldValidateRules": [
                "required"
            ]
        },
        {
            "fieldName": "cohort",
            "fieldType": "String"
        },
        {
            "fieldName": "ethnicity",
            "fieldType": "String"
        },
        {
            "fieldName": "gender",
            "fieldType": "GenderType",
            "fieldValues": "female,male,unknown",
            "fieldValidateRules": [
                "required"
            ]
        },
        {
            "fieldName": "primary_diagnosis",
            "fieldType": "String"
        }
    ],
    "changelogDate": "***",
    "dto": "no",
    "service": "no",
    "entityTableName": "patient_info",
    "pagination": "pagination"
}

```

##### **Browsers and Operating System**

java version "1.8.0_60"
Java(TM) SE Runtime Environment (build 1.8.0_60-b27)
Java HotSpot(TM) 64-Bit Server VM (build 25.60-b23, mixed mode)

git version 2.6.4 (Apple Git-63)

node: v6.9.5

npm: 3.10.10

bower: 1.8.0

gulp:
[10:31:12] CLI version 1.2.2
[10:31:12] Local version 3.9.1

yeoman: 1.8.5

yarn: 0.21.3

Docker version 17.03.0-ce, build 60ccb22

docker-compose version 1.11.2, build dfed245

### Output from creating a new row

```
curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' --header 'X-XSRF-TOKEN: ***' -d '{ \ 
   "cohort": "string", \ 
   "ethnicity": "string", \ 
   "gender": "female", \ 
   "patient_id": "PAT002", \ 
   "primary_diagnosis": "string" \ 
 }' 'http://localhost:8080/api/patient-infos'
```

Response body:
```
{
  "message": "error.internalServerError",
  "description": "Internal server error",
  "fieldErrors": null
}
```

Response Code: 500

Response Header:
```
{
  "content-encoding": "gzip",
  "expires": "0",
  "cache-control": "no-cache, no-store, max-age=0, must-revalidate",
  "x-xss-protection": "1; mode=block",
  "pragma": "no-cache",
  "date": "Mon, 27 Mar 2017 21:25:38 GMT",
  "connection": "keep-alive",
  "x-content-type-options": "nosniff",
  "transfer-encoding": "chunked",
  "content-type": "application/json;charset=UTF-8",
  "x-application-context": "mysql_test:prod,swagger:8080"
}
```

**Comment by pascalgrimaud** [comment]
I didn't manage to reproduce.
Do you have additional information, on how to reproduce your issue?

Here all my steps
- I generated a new application with your `.yo-rc.json`, adding a proper `rememberMeKey`
- I generated your entity
- start docker mysql and elasticsearch
- `jhpack` to build in prod
- run with `java -jar target/*.war --spring.profiles.active=prod,swagger`

Then:

```shell
curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' --header 'X-XSRF-TOKEN: 8decde79-79f6-441d-9ff2-c9a73c97084b' -d '{ \ 
    "cohort": "string", \ 
    "ethnicity": "string", \ 
    "gender": "female", \ 
    "patient_id": "PAT002", \ 
    "primary_diagnosis": "string" \ 
  }' 'http://localhost:8080/api/patient-infos'
```

Request URL:
```
http://localhost:8080/api/patient-infos
```

Request body:
```yaml
{
  "id": 1,
  "patient_id": "PAT002",
  "cohort": "string",
  "ethnicity": "string",
  "gender": "female",
  "primary_diagnosis": "string"
}
```

Response code:
```
201
```

Response Headers:
```yaml
{
  "pragma": "no-cache",
  "date": "Tue, 28 Mar 2017 06:51:40 GMT",
  "content-encoding": "gzip",
  "x-content-type-options": "nosniff",
  "x-mysqltestapp-alert": "A new patientInfo is created with identifier 1",
  "transfer-encoding": "chunked",
  "content-type": "application/json;charset=UTF-8",
  "location": "/api/patient-infos/1",
  "cache-control": "no-cache, no-store, max-age=0, must-revalidate",
  "connection": "keep-alive",
  "x-mysqltestapp-params": "1",
  "x-xss-protection": "1; mode=block",
  "x-application-context": "mysql_test:prod,swagger:8080",
  "expires": "0"
}
```

**Comment by jdubois** [comment]
Yes @avansp if you have an internal server error, we need your stacktrace.
Here we can just guess at what you are doing. I'm closing this as we don't have enough information for the moment.

**Comment by avansp** [comment]
@pascalgrimaud, I think the error happens when I used a remote MySQL server.

I did using docker like what you did and it went ok with response code = 200.

And here I tried using a remote MySQL server running on a VM on my local machine:

Settings: MySQL 5.7.14, user: cap with all privileges on `*.*` with grant option.

Steps:

1. Changed `application-prod.yml` appropriately:

```
url: jdbc:mysql://192.168.56.201:3306/mysql_test?useUnicode=true&characterEncoding=utf8&useSSL=false
username: cap
password: ***
```

2. Built the app: `./mvnw -Pprod,swagger package`

3. Ran it: `java -jar target/*.war`

4. Go to swagger API and call the create patient-info using:

```
{
  "patient_id": "PAT002",
  "cohort": "string",
  "ethnicity": "string",
  "gender": "female",
  "primary_diagnosis": "string"
}
```

  RESPONSE BODY:
```
{
  "message": "error.internalServerError",
  "description": "Internal server error",
  "fieldErrors": null
}
```

  RESPONSE: 500
  
  RESPONSE HEADER:
  ```
{
  "content-encoding": "gzip",
  "expires": "0",
  "cache-control": "no-cache, no-store, max-age=0, must-revalidate",
  "x-xss-protection": "1; mode=block",
  "pragma": "no-cache",
  "date": "Tue, 28 Mar 2017 21:53:41 GMT",
  "connection": "keep-alive",
  "x-content-type-options": "nosniff",
  "transfer-encoding": "chunked",
  "content-type": "application/json;charset=UTF-8",
  "x-application-context": "mysql_test:prod,swagger:8080"
}
  ```

**Comment by avansp** [comment]
Just to continue, the **transaction is complete** in the database after that error.

```
curl -X GET --header 'Accept: application/json' --header 'X-XSRF-TOKEN: 157d1d04-e811-4f45-b0e9-02d0bfd84bb1' 'http://localhost:8080/api/patient-infos'
```

Response body:
```
[
  {
    "id": 1,
    "patient_id": "PAT002",
    "cohort": "string",
    "ethnicity": "string",
    "gender": "female",
    "primary_diagnosis": "string"
  }
]
```

Response code: 200

Response header:
```
{
  "content-encoding": "gzip",
  "expires": "0",
  "cache-control": "no-cache, no-store, max-age=0, must-revalidate",
  "x-xss-protection": "1; mode=block",
  "pragma": "no-cache",
  "date": "Tue, 28 Mar 2017 22:00:33 GMT",
  "x-total-count": "1",
  "connection": "keep-alive",
  "x-content-type-options": "nosniff",
  "transfer-encoding": "chunked",
  "content-type": "application/json;charset=UTF-8",
  "x-application-context": "mysql_test:prod,swagger:8080",
  "link": "</api/patient-infos?page=0&size=20>; rel=\"last\",</api/patient-infos?page=0&size=20>; rel=\"first\""
}
```

**Comment by ruddell** [comment]
What do your Java logs say?  We need that stacktrace

**Comment by avansp** [comment]
Where can I find the java logs for the production mode?

**Comment by ruddell** [comment]
In the terminal where you ran `java -jar target/*.war`

**Comment by ruddell** [comment]
Your issue happens because you didn't configure elasticsearch in prod.  Without any available nodes, the entity's save method fails (it saves to elasticsearchh in that method).

>By default, we use an embedded Elasticsearch in development, and try to connect to a local cluster in production. This can be configured by the standard Spring Boot properties, in the application.yml file. (see the [elasticsearch docs](https://jhipster.github.io/using-elasticsearch/))

It is weird that the transaction completes even though there is an error

**Comment by avansp** [comment]
Thanks !! I'll remove elasticsearch from my app.

**Comment by jdubois** [comment]
You misconfigured MySQL and didn't even configure Elasticsearch... I'm sorry, but there is definitely no JHipster bug here!

**Comment by avansp** [comment]
How come the transaction complete though?

**Comment by jdubois** [comment]
It didn't - I mean, you didn't even set up your database properly, I'm sure your data isn't there. What probably happened is this:

- The database transaction failed (have a look in your database table)
- You didn't configure Elasticsearch, so JHipster started an in-memory instance (that's the default with Spring Data Elasticsearch -> have a look in your logs, or in your "admin" pages, you'll see this)
- So your data got inserted into this temporary Elasticsearch instance, but not the database
- That's why you got an error (from the database), but the table could show your data (as it was read from Elasticsearch)

Then I really don't have enough information on what you did, so here I'm just guessing.

**Comment by avansp** [comment]
Well, yes it did. I checked my mysql database directly and the data got inserted.
Anyway, I've removed elasticsearch, so it doesn't matter for me anymore.

**Comment by rhanberry** [comment]
This was informative for me as I was experiencing a similar problem (albeit with Postgresql). As a solution perhaps someone (I would be willing to do this once I have my app fixed and configured correctly, but I'll have to learn) could update the documentation to reflect production configuration of elasticsearch, as it isn't immediately intuitive (unless I am missing something, which is likely)

**Comment by ruddell** [comment]
@rhanberry Is something missing from the documentation?  This does seem to be a common issue but the docs mention it.  I'm not sure of a better default setting either.

>By default, we use an embedded Elasticsearch in development, and try to connect to a local cluster in production. This can be configured by the standard Spring Boot properties, in the application.yml file.
https://jhipster.github.io/using-elasticsearch/

**Comment by gmarziou** [comment]
@ruddel you're right this a frequent error. Maybe there is a way to stop the application if elasticsearch cannot be reached at startup, this would make it more obvious that ES is the issue.

**Comment by jdubois** [comment]
I'm not saying that's a great design, but that's how Spring Boot works by default, see [here](https://github.com/spring-projects/spring-boot/blob/v1.5.2.RELEASE/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/elasticsearch/ElasticsearchProperties.java#L40).
So reconfiguring this, not to start the embedded Elasticsearch node, would be against [policy 1](https://jhipster.github.io/policies/)

**Comment by gmarziou** [comment]
Right and as [spring-data catch the exception](https://github.com/spring-projects/spring-data-elasticsearch/blob/master/src/main/java/org/springframework/data/elasticsearch/repository/support/AbstractElasticsearchRepository.java#L80) there's nothing we can do.

**Comment by rhanberry** [comment]
I just think it might be more clear if it was explicitly stated that we are on our own for establishing an elasticsearch server in production, even locally it's still a few extra steps to get that going on a prod server.  
It wasn't immediately clear to me when I added elasticsearch to my app, and it would have saved me a lot of time if I had known that I needed to do that from the outset for my production deployment. 
Not required by any stretch, Jhipster is so nice in that everything is pretty functional right out of the box, but I brought it up because it did cause me a bit of futzing around when I encountered this issue.
